test/src/tests-session-manager.ts

```diff
@@ -0,0 +1,607 @@
+                },
+            });
+
+            mpInstance._SessionManager.endSession();
+
+            expect(mpInstance._Store.sessionId).to.equal('fake-session-id');
+
+            // When session is not timed out, setSessionTimer is called to keep track
+            // of current session timeout
+            expect(timerSpy.getCalls().length).to.equal(1);
+        });
+
+        it('should end session if session times out', () => {
+            const now = new Date();
+            const hourAgo = new Date();
+            hourAgo.setMinutes(now.getMinutes() - 60);
+
+            mParticle.init(apiKey, window.mParticle.config);
+            const mpInstance = mParticle.getInstance();
+            const persistenceSpy = sinon.spy(mpInstance._Persistence, 'update');
+
+            // Session Manager relies on persistence to determine last time seen (LES)
+            // Also requires sid to verify session exists
+            sinon.stub(mpInstance._Persistence, 'getPersistence').returns({
+                gs: {
+                    les: hourAgo,
+                    sid: 'fake-session-id',
+                },
+            });
+
+            mpInstance._SessionManager.endSession(true);
+
+            expect(mpInstance._Store.sessionId).to.equal(null);
+            expect(mpInstance._Store.dateLastEventSent).to.equal(null);
+            expect(mpInstance._Store.sessionAttributes).to.eql({});
+
+            // Persistence isn't necessary for this feature, but we should test
+            // to see that it is called in case this ever needs to be refactored
+            expect(persistenceSpy.called).to.equal(true);
+        });
+    });
+
+    describe('#setSessionTimer', () => {
+        it('should end a session after the timeout expires', done => {
+            mParticle.init(apiKey, window.mParticle.config);
+            const mpInstance = mParticle.getInstance();
+            const endSessionSpy = sinon.spy(
+                mpInstance._SessionManager,
+                'endSession'
+            );
+
+            // Start Timer
+            mpInstance._SessionManager.setSessionTimer();
+
+            // Progress 29 minutes to make sure end session has not fired
+            clock.tick(29 * 60000);
+            expect(endSessionSpy.called).to.equal(false);
+
+            // Progress one minutes to make sure end session fires
+            clock.tick(30 * 60000);
+            expect(endSessionSpy.called).to.equal(true);
+
+            // Let test know async process is done
+            done();
+        });
+
+        it('should set a global timer', () => {
+            mParticle.init(apiKey, window.mParticle.config);
+            const mpInstance = mParticle.getInstance();
+
+            // Manually clear global timer since it gets set during init
+            // window.clearTimeout(mpInstance._Store.globalTimer);
+            mpInstance._Store.globalTimer = null;
+            expect(mpInstance._Store.globalTimer).to.eq(null);
```

Comment:
Should line 375 be uncommented, and should line 376 be deleted?

The way I see it, 376 sets a null, then 377 checks if it's null, which is a bit superfluous.  If you uncomment line 375 and delete 376, does the test still pass?

Reply:
In my testing, `window.clearTimeout(mpInstance._Store.globalTimer);` doesn't actually clear out the timer. I think that's why I commented it out in the first place and forgot to delete it.